python/ray/tune/tests/test_tuner_restore.py

```diff
@@ -746,6 +748,56 @@ def test_restore_from_relative_path(ray_start_4_cpus, chdir_tmpdir):
     assert results[0].metrics["score"] == 1
 
 
+def test_custom_searcher_and_scheduler_restore(ray_start_2_cpus, tmpdir):
+    """Check that a restored Tune experiment uses the original searcher/scheduler."""
+    fail_marker = tmpdir / "fail_marker"
+    fail_marker.write_text("", encoding="utf-8")
+
+    class MockSearcher(OptunaSearch):
+        def on_trial_result(self, trial_id: str, result: dict):
+            super().on_trial_result(trial_id, result)
+            if not hasattr(self, "_test_result_counter"):
+                self._test_result_counter = 0
+            self._test_result_counter += 1
+
+    class MockScheduler(ASHAScheduler):
+        def on_trial_result(self, runner, trial, result):
+            decision = super().on_trial_result(runner, trial, result)
+            if not hasattr(self, "_test_result_counter"):
+                self._test_result_counter = 0
+            self._test_result_counter += 1
+            return decision
+
+    tuner = Tuner(
+        _train_fn_sometimes_failing,
+        run_config=RunConfig(local_dir=str(tmpdir), name="exp_name"),
+        tune_config=TuneConfig(
+            search_alg=MockSearcher(),
+            scheduler=MockScheduler(),
+            metric="it",
+            mode="max",
+        ),
+        param_space={"a": tune.uniform(0, 1), "failing_hanging": (fail_marker, None)},
+    )
+    tuner.fit()
+
+    del tuner
+    fail_marker.remove(ignore_errors=True)
+
+    tuner = Tuner.restore(str(tmpdir / "exp_name"), resume_errored=True)
+    tuner.fit()
+    searcher = tuner._local_tuner._tune_config.search_alg
+    scheduler = tuner._local_tuner._tune_config.scheduler
+    # Searcher state should get loaded correctly
+    # Total of 3 reported results (1 from before failure, 2 after restore)
```

Comment:
Nit (imo introducing these assertions makes the test a bit more readable)

```suggestion
    searcher = tuner._local_tuner._tune_config.search_alg
    scheduler = tuner._local_tuner._tune_config.scheduler
    assert isinstance(searcher, MockSearcher)
    assert isinstance(scheduler, MockScheduler)
    # Searcher state should get loaded correctly
    # Total of 3 reported results (1 from before failure, 2 after restore)
```